### PR TITLE
Universal Packages

### PR DESCRIPTION
**Pull request by freddydk**
Proof of Concept support for hosting artifacts as Universal Packages

### CODE REVIEW

File: Artifacts/Get-BCArtifactUrl.ps1 [hunk 1]
```diff
@@ -188,99 +192,124 @@ try {
...
-        $downloadRetryAttempts = 10
-        do {
-            if ($currentMarker -ne $nextMarker)
-            {
-                $currentMarker = $nextMarker
-                $downloadAttempt = 1
+
+        if ($bcContainerHelperConfig.useUniversalPackagesForArtifacts) {
+            $upArr = $bcContainerHelperConfig.useUniversalPackagesForArtifacts.Split('/')
+            $feedApiUrl = "https://feeds.dev.azure.com/$($upArr[0])/$($upArr[1])/_apis/packaging/feeds/$($storageAccount.Split('.')[0])"
```
Comment (mazhelez) [inline Artifacts/Get-BCArtifactUrl.ps1:201]: Split by `/` only to join by `/`? Some code comments around here would be useful.

File: Artifacts/Get-BCArtifactUrl.ps1 [hunk 2]
```diff
@@ -188,99 +192,124 @@ try {
...
-            Write-Verbose "Download String $GetListUrl$nextMarker"
-            try
-            {
-                $Response = Invoke-RestMethod -UseBasicParsing -ContentType "application/json; charset=UTF8" -Uri "$GetListUrl$nextMarker"
-                if (([int]$Response[0]) -eq 239 -and ([int]$Response[1]) -eq 187 -and ([int]$Response[2]) -eq 191) {
-                    # Remove UTF8 BOM
-                    $response = $response.Substring(3)
+            $result = invoke-restmethod -UseBasicParsing -Uri "$feedApiUrl/packages?api-version=7.0$query&includeAllVersions=$(($select -ne 'latest').ToString().ToLowerInvariant())"
+            $Artifacts = @($result.value | ForEach-Object {
+                $nameArr = $_.name.Split('.')
```
Comment (mazhelez) [inline Artifacts/Get-BCArtifactUrl.ps1:211]: What is the format for each `$result.value`?

**Review by mazhelez — COMMENTED** [review]
The format of the new `useUniversalPackagesForArtifacts` isn't clear.

As well as the new packages names. See comments.

File: Artifacts/Get-BCArtifactUrl.ps1 (hunk 1 above)
Comment (freddydk) [reply Artifacts/Get-BCArtifactUrl.ps1:201]: :-)

File: Artifacts/Download-Artifacts.ps1 [hunk 3]
```diff
@@ -33,6 +33,80 @@ function Download-Artifacts {
...
+    function DownloadPackage {
+        Param(
+            [string] $artifactUrl,
+            [string] $destinationPath,
+            [int]    $timeout = 300
+        )
+
+        $tmpFolder = Join-Path ([System.IO.Path]::GetDirectoryName($destinationPath)) ([System.IO.Path]::GetRandomFileName())
+        $zipFile = Join-Path ([System.IO.Path]::GetTempPath()) "$([Guid]::NewGuid().ToString()).zip"
+        if ($bcContainerHelperConfig.UseUniversalPackagesForArtifacts -and ($artifactUrl -match '^https:\/\/([^.]+)\..*\.net\/([^\/]+)\/([^\/]+)\/([a-zA-Z0-9]+).*$')) {
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Artifacts/Download-Artifacts.ps1]: agree

File: Artifacts/Download-Artifacts.ps1 [hunk 4]
```diff
@@ -33,6 +33,80 @@ function Download-Artifacts {
...
+            $upArr = $bcContainerHelperConfig.useUniversalPackagesForArtifacts.Split('/')
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Artifacts/Download-Artifacts.ps1]: changed name of setting

File: Artifacts/Download-Artifacts.ps1 [hunk 5]
```diff
@@ -33,6 +33,80 @@ function Download-Artifacts {
...
+        try {
+            while (!(Test-Path $destinationPath)) {
+                try {
+                    Rename-Item -Path $tmpFolder -NewName ([System.IO.Path]::GetFileName($destinationPath)) -Force
+                    $result = $true
+                }
+                catch {
+                    Write-Host "Could not rename '$tmpFolder' retrying in 5 seconds."
+                    Start-Sleep -Seconds 5
+                    Write-Host "Retrying..."
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Artifacts/Download-Artifacts.ps1:99]: Correct - this actually haven't changed - same behavior has been there for a very long time.
It is primarily a problem from unzipping might keep a handle in some files for a short while - but I cannot say whether it is 5, 10 or 15 seconds.
I will add a counter though

File: Artifacts/Download-Artifacts.ps1 [hunk 6]
```diff
@@ -33,6 +33,90 @@ function Download-Artifacts {
...
+                Download-File -sourceUrl $artifactUrl -destinationFile $zipFile -timeout $timeout
+                Write-Host "Unpacking artifact to tmp folder " -NoNewline
+                try {
+                    Expand-7zipArchive -Path $zipFile -DestinationPath $tmpFolder -use7zipIfAvailable:(!$retry)
+                    $retry = $false
+                }
+                catch {
+                    Remove-Item -path $zipFile -force
+                    if (Test-Path $tmpFolder) {
+                        Remove-Item $tmpFolder -Recurse -Force
```
Comment (mazhelez) [inline Artifacts/Download-Artifacts.ps1:74]: Isn't there a case when `$tmpFolder` has been deleted, but the there is a retry? Will there be side effects from this?

**Review by mazhelez — APPROVED** [review]
Code comments were quite helpful when reviewing, thanks!

File: Artifacts/Download-Artifacts.ps1 (hunk 6 above)
Comment (freddydk) [reply Artifacts/Download-Artifacts.ps1:74]: This is the "old" code - expand-7zip creates the folder, so it needs to be deleted before retrying

**Review by mazhelez — APPROVED** [review]

**Pull request merged**